gpcontrib/arenadata_toolkit/src/track_files.c

```diff
@@ -0,0 +1,1353 @@
+
+	rel = heap_open(DbRoleSettingRelationId, RowExclusiveLock);
+	ScanKeyInit(&skey[0],
+				Anum_pg_db_role_setting_setdatabase,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(MyDatabaseId));
+
+	/*
+	 * Lookup for not role specific configuration
+	 */
+	ScanKeyInit(&skey[1],
+				Anum_pg_db_role_setting_setrole,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(InvalidOid));
+	scan = systable_beginscan(rel, DbRoleSettingDatidRolidIndexId, true, NULL, 2, skey);
+
+	tuple = systable_getnext(scan);
+	if (HeapTupleIsValid(tuple))
+	{
+		bool		isnull;
+		Datum		str_datum;
+
+		str_datum = heap_getattr(tuple, Anum_pg_db_role_setting_setconfig,
+								 RelationGetDescr(rel), &isnull);
+		if (!isnull)
+		{
+			ArrayType  *array;
+			Datum	   *elems;
+			bool	   *nulls;
+			int			nelems;
+
+			array = DatumGetArrayTypeP(str_datum);
+			deconstruct_array(array, TEXTOID, -1, false, 'i',
+							  &elems, &nulls, &nelems);
+			for (int i = 0; i < nelems; i++)
+			{
+				if (nulls[i])
+					continue;
+				char	   *str = TextDatumGetCString(elems[i]);
+
+				if (strncmp(str, "arenadata_toolkit.tracking_schemas=", 35) == 0)
+					current_schemas = pstrdup(str + 35);
+				else if (strncmp(str, "arenadata_toolkit.tracking_relstorages=", 39) == 0)
+					current_relstorages = pstrdup(str + 39);
+				else if (strncmp(str, "arenadata_toolkit.tracking_relkinds=", 36) == 0)
+					current_relkinds = pstrdup(str + 36);
+				pfree(str);
+			}
+		}
+	}
+	systable_endscan(scan);
+	heap_close(rel, RowExclusiveLock);
+
+	if (current_schemas)
+		schema_names = split_string_to_list(current_schemas);
+	else
+		schema_names = split_string_to_list(tracked_schemas);
+	if (current_relstorages)
+		tf_get_global_state.relstorages = list_to_bits(current_relstorages);
+	else
+		tf_get_global_state.relstorages = list_to_bits(tracked_rel_storages);
+	if (current_relkinds)
+		tf_get_global_state.relkinds = list_to_bits(current_relkinds);
+	else
+		tf_get_global_state.relkinds = list_to_bits(tracked_rel_kinds);
+
+	foreach(lc, schema_names)
+	{
+		Oid			nspOid;
+		char	   *name = (char *) lfirst(lc);
+
+		nspOid = GetSysCacheOid1(NAMESPACENAME, CStringGetDatum(name));
+
+		if (!OidIsValid(nspOid))
+		{
+			elog(DEBUG1, "[tracking_get_track] schema \"%s\" does not exist", name);
+			continue;
+		}
+
+		tf_get_global_state.schema_oids = lappend_oid(tf_get_global_state.schema_oids, nspOid);
+	}
+
+	if (schema_names)
+		pfree(schema_names);
+}
+
+
+static bool
+schema_is_tracked(Oid schema)
+{
+	ListCell   *lc;
+
+	if (tf_get_global_state.schema_oids == NIL)
+		return false;
+
+	foreach(lc, tf_get_global_state.schema_oids)
+	{
+		Oid			tracked_schema = lfirst_oid(lc);
+
+		if (tracked_schema == schema)
+			return true;
+	}
+
+	return false;
+}
+
+static bool
+kind_is_tracked(char type, uint64 allowed_kinds)
+{
+	if (type >= 'a' && type <= 'z')
+		return (allowed_kinds & (1UL << (type - 'a'))) != 0;
+	else
+		return (allowed_kinds & (1UL << (26 + (type - 'A')))) != 0;
```

Comment:
```suggestion
	return (allowed_kinds & (1UL << (type - 'A'))) != 0;
```
I think we can just do `type - 'A'`. It will waste some bits between a-z and A-Z, but we still should fit into uint64.